DRILL-7403: Validate batch checks, vector integretity in unit tests

paul-rogers opened the pull request:
Enhances the existing record batch checks to check all the various
batch record counts, and to more fully validate all vector types.

This code revealed that virtually all record batches have
problems: they omit setting some record count or other, they
introduce some form of vector corruption.

Since we want things to work as we make fixes, this change enables
the checks for only one record batch: the "new" scan. Others are
to come as they are fixed.

Jira - [DRILL-7403](https://issues.apache.org/jira/browse/DRILL-7403).

paul-rogers commented:
@arina-ielchiieva, please review. As we discussed, the idea here is to add the "vector checker", but with it disabled for most operators. Then, we can gradually add the fixes from that earlier "record count" PR, enabling checks as we go. Should make for smaller, easier-to-review and test PRs.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/validate/BatchValidator.java:
```diff
@@ -101,47 +290,90 @@ private void validateVector(ValueVector vector) {
+  private int validateOffsetVector(String name, UInt4Vector offsetVector, boolean repeated, int valueCount, int maxOffset) {
+    UInt4Vector.Accessor accessor = offsetVector.getAccessor();
+    int offsetCount = accessor.getValueCount();
+    // Disabled because a large number of operators
```
Should we leave this comment and commented code?

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/validate/IteratorValidatorBatchIterator.java:
```diff
@@ -42,15 +42,15 @@
   private static final org.slf4j.Logger logger =
```
Imports :)

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/validate/IteratorValidatorBatchIterator.java:
```diff
@@ -342,8 +342,15 @@ public IterOutcome next() {
+            "Batch validation failed. Source operator: " +
+            incoming.getClass().getSimpleName());
+      }
+      // The above seems to validate hash codes, but just calculates
```
Do we need this comment?

paul-rogers commented:
@arina-ielchiieva, addressed the comments. Since they were minor, went ahead and squashed commits.

arina-ielchiieva commented:
LGTM, +1

paul-rogers commented:
Fixed some issues for repeated types and the bit vector.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/validate/BatchValidator.java:
```diff
@@ -45,103 +49,373 @@
   }
 
-  private void validateRepeatedVector(String name, BaseRepeatedValueVector vector) {
+//  private void validateRepeatedBitVector(String name, BaseRepeatedValueVector vector) {
```
Please remove.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/validate/BatchValidator.java:
```diff
@@ -150,57 +424,63 @@ private int validateOffsetVector(String name, UInt4Vector offsetVector, int valu
+    for (int i = 0; i < bitCount; i++) {
+      final int value = ba.get(i);
+      if (value != 0 && value != 1) {
+        error(name, bv, String.format(
```
Looks like string format expecting 4 parameters, but only two are passed.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/validate/BatchValidator.java:
```diff
@@ -150,57 +424,63 @@ private int validateOffsetVector(String name, UInt4Vector offsetVector, int valu
-  public List<String> errors() { return errorList; }
+  public static void print(RecordBatch batch) {
+    try {
+      final Class<?> helper = Class.forName("org.apache.drill.test.rowSet.RowSetUtilities");
```
Checked `org.apache.drill.test.rowSet.RowSetUtilities` it does not have `print` method. Could you please check?

arina-ielchiieva commented:
@paul-rogers sorry could not merge this commit, since there a couple of minor comments. Mostly I am worried about absent `print` method.

paul-rogers commented:
Made requested changes. Rebased on master. Squashed commits.

Once this is merged, I'll update the two new PRs to eliminate the commits duplicated with this one.

arina-ielchiieva commented:
+1
